landoapi/workers/base.py

```diff
@@ -0,0 +1,136 @@
+        push path.
+
+        The private key should be passed as it is in the key file, including all
+        new line characters and the new line character at the end.
+
+        Args:
+            ssh_private_key (str): A string representing the private SSH key file.
+        """
+        # Set all the correct environment variables
+        agent_process = subprocess.run(
+            ["ssh-agent", "-s"], capture_output=True, universal_newlines=True
+        )
+
+        # This pattern will match keys and values, and ignore everything after the
+        # semicolon. For example, the output of `agent_process` is of the form:
+        #     SSH_AUTH_SOCK=/tmp/ssh-c850kLXXOS5e/agent.120801; export SSH_AUTH_SOCK;
+        #     SSH_AGENT_PID=120802; export SSH_AGENT_PID;
+        #     echo Agent pid 120802;
+        pattern = re.compile("(.+)=([^;]*)")
+        for key, value in pattern.findall(agent_process.stdout):
+            logger.info(f"_setup_ssh: setting {key} to {value}")
+            os.environ[key] = value
+
+        # Add private SSH key to agent
+        # NOTE: ssh-add seems to output everything to stderr, including upon exit 0.
+        add_process = subprocess.run(
+            ["ssh-add", "-"],
+            input=ssh_private_key,
+            capture_output=True,
+            universal_newlines=True,
+        )
+        if add_process.returncode != 0:
+            raise Exception(add_process.stderr)
+        logger.info("Added private SSH key from environment.")
+
+    @property
+    def _paused(self):
+        return ConfigurationVariable.get(self.PAUSE_KEY, False)
+
+    @property
+    def _running(self):
+        # When STOP_KEY is False, the worker is running.
+        return not ConfigurationVariable.get(self.STOP_KEY, False)
+
+    def _setup(self):
+        if hasattr(self, "ssh_private_key"):
+            self._setup_ssh(self.ssh_private_key)
+
+    def _start(self, max_loops=None, *args, **kwargs):
+        loops = 0
+        while self._running:
+            if max_loops is not None and loops >= max_loops:
+                break
+            while self._paused:
+                self.throttle(self.sleep_seconds)
+            self.loop(*args, **kwargs)
+            loops += 1
+
+        logger.info(f"{self} exited after {loops} loops.")
+
+    @property
+    def throttle_seconds(self):
+        return ConfigurationVariable.get(self.THROTTLE_KEY, 3)
+
+    def throttle(self, seconds=None):
+        sleep(seconds if seconds is not None else self.throttle_seconds)
+
+    def refresh_enabled_repos(self):
+        self.enabled_repos = [
+            r
+            for r in self.applicable_repos
+            if treestatus_subsystem.client.is_open(repo_clone_subsystem.repos[r].tree)
+        ]
+        logger.info(f"{len(self.enabled_repos)} enabled repos: {self.enabled_repos}")
+
+    def start(self, max_loops=None):
+        if ConfigurationVariable.get(self.STOP_KEY, False):
```

Comment:
There is no `START_KEY`, `STOP_KEY` is used to determine whether a worker can be started or not (as seen in this method).

I don't see a harm in moving this to a property as you described, since this is an abstract class and it would make sense to have unimplemented properties that should exist in all extended class here. This should probably be upgraded to a proper abstract base class at some point.